oauth2client/service_account.py

```diff
@@ -122,6 +122,81 @@ def create_scoped(self, scopes):
                                       **self._kwargs)
 
 
+class _ServiceAccountJwtHeaderCredentials(AssertionCredentials):
+  """Class representing a service account (signed JWT) credential."""
+
+  MAX_TOKEN_LIFETIME_SECS = 3600 # 1 hour in seconds
+
+  def __init__(self, service_account_id, service_account_email, private_key_id,
+               private_key_pkcs8_text, user_agent=None,
+               token_uri=GOOGLE_TOKEN_URI, revoke_uri=GOOGLE_REVOKE_URI,
+               **kwargs):
+
+    super(_ServiceAccountJwtHeaderCredentials, self).__init__(
+        None, user_agent=user_agent, token_uri=token_uri, revoke_uri=revoke_uri)
+
+    self._service_account_id = service_account_id
+    self._service_account_email = service_account_email
+    self._private_key_id = private_key_id
+    self._private_key = _get_private_key(private_key_pkcs8_text)
+    self._private_key_pkcs8_text = private_key_pkcs8_text
+    self._user_agent = user_agent
+    self._token_uri = token_uri
+    self._revoke_uri = revoke_uri
+    self._kwargs = kwargs
+
+  def _generate_assertion(self):
+    """Generate the assertion that will be used in the request."""
+
+    now = int(time.time())
+    jwt_token = {
+        'aud': self._token_uri,
+        'iat': now,
+        'exp': now + _ServiceAccountCredentials.MAX_TOKEN_LIFETIME_SECS,
+        'iss': self._service_account_email,
+        'sub': self._service_account_email
+    }
+    signed_jwt_token = self.sign_blob(_urlsafe_b64encode(jwt_token))
```

Comment:
What is the proper order of token -> encode -> sign -> place-in-header -> encode?

Currently the encoding procedure complains about invalid start bytes, and I imagine that has to do with signing making the token an arbitrary byte string rather than proper unicode.